src/compiler/checker.ts

```diff
@@ -27797,7 +27797,8 @@ namespace ts {
             // The in operator requires the left operand to be of type Any, the String primitive type, or the Number primitive type,
             // and the right operand to be of type Any, an object type, or a type parameter type.
             // The result is always of the Boolean primitive type.
-            if (!(isTypeComparableTo(leftType, stringType) || isTypeAssignableToKind(leftType, TypeFlags.NumberLike | TypeFlags.ESSymbolLike))) {
+            if (!allTypesAssignableToKind(leftType, TypeFlags.StringLike | TypeFlags.NumberLike | TypeFlags.ESSymbolLike) &&

```

Comment:
Two comments:

1. I had to mentally convert your `(!A && !(B && C))` to `(!(A || (B && C)))` to follow the conditions, so I think that the latter is more readable.
2. Is there a need for the `isTypeComparableTo` test?  I tried dropping it, and passed the tests.

```suggestion
            if (!(allTypesAssignableToKind(leftType, TypeFlags.StringLike | TypeFlags.NumberLike | TypeFlags.ESSymbolLike) ||
                  isTypeAssignableToKind(leftType, TypeFlags.Index | TypeFlags.TypeParameter))) {
```

Reply:
@elibarzilay Thanks for the feedback. Fixed.